packages/crud/test/crud.test.js

```diff
@@ -121,35 +121,78 @@ describe('crud', () => {
   describe('dataProvider', () => {
     const items = [{ foo: 'bar' }];
 
-    beforeEach(async () => {
-      crud = fixtureSync('<vaadin-crud style="width: 300px;"></vaadin-crud>');
-      crud.dataProvider = (_, callback) => callback(items, items.length);
-      await nextRender(crud);
-      [btnSave, btnCancel, btnDelete] = crud.querySelectorAll('vaadin-button');
-    });
+    describe('passing to grid', () => {
+      let customGrid;
 
-    it('should save a new item', (done) => {
-      listenOnce(crud, 'save', (e) => {
-        expect(e.detail.item.foo).to.be.equal('baz');
-        done();
+      beforeEach(() => {
+        crud = document.createElement('vaadin-crud');
+        customGrid = fixtureSync(`
+          <vaadin-grid>
+            <vaadin-grid-column path="foo"></vaadin-grid-column>
+          </vaadin-grid>
+        `);
+        customGrid.setAttribute('slot', 'grid');
+      });
+
+      afterEach(() => {
+        crud.remove();
+      });
+
+      it('should apply data provider when setting data provider before initializing the grid', async () => {
+        // Set data provider before `ready` is called, which means the grid is not initialized yet
+        crud.dataProvider = (_, callback) => callback(items, items.length);
+        // This should initialize the grid and pass the data provider
+        document.body.appendChild(crud);
+        // Verify item from data provider is rendered
+        await nextRender(crud);
+        flushGrid(crud._grid);
+        expect(getBodyCellContent(crud._grid, 0, 0).textContent).to.be.equal('bar');
+      });
```

Comment:
I had a thought that we might've added also a test verifying that the data provider is applied also in the case of using a custom grid (not replacing as below) but is probably too much and there is not any point in it. Leaving this here just for the record.

Reply:
Added a test for that as well.